embassy-stm32/src/rcc/u5.rs

```diff
@@ -195,24 +228,38 @@ pub(crate) unsafe fn init(config: Config) {
         RCC.cr().modify(|w| {
             w.set_msikon(true);
         });
-        if lse_calibration_freq.is_some() {
+        let msik = if let (Some(freq), MsiAutoCalibration::MSIK) = (lse_calibration_freq, config.auto_calibration) {
             // Enable the MSIK auto-calibration feature
             RCC.cr().modify(|w| w.set_msipllsel(Msipllsel::MSIK));
             RCC.cr().modify(|w| w.set_msipllen(true));
-        }
-        while !RCC.cr().read().msikrdy() {}
-        if let Some(freq) = lse_calibration_freq {
-            let msik_freq = calculate_calibrated_msi_frequency(range, freq);
-            if config.msis == config.msik {
-                // If MSIS and MSIK are the same range both will be auto calibrated to the same frequency
-                msis = Some(msik_freq)
-            }
-            msik_freq
+            calculate_calibrated_msi_frequency(range, freq)
         } else {
             msirange_to_hertz(range)
-        }
+        };
+        while !RCC.cr().read().msikrdy() {}
+        msik
     });
 
+    // If both MSIS and MSIK are enabled, we need to check if they are using the same internal source.
+    if let Some(lse_freq) = lse_calibration_freq {
+        if let (Some(msis_range), Some(msik_range)) = (config.msis, config.msik) {
```

Comment:
this makes autocalibration only be applied when both msis and msik are enabled, which I think is too restrictive?

for example if config has `msis=Some(..), msik=None, autocalbration=MSIS`, autocalibration will not be applied, which would be a bit surprising.

We should also probably panic if the user has set autocalbration=MSIx and the corresponding MSIx is None, that's an invalid config and we shouldn't silently ignore it.

Reply:
I can add a panic!() - I was just hesitant to do so...